### PR TITLE
Readonly via expressions

### LINKED ISSUE #396: Support `readonly` fields via FEEL expression
__What should we do?__

Implement the `readonly` property.

Sketches
image

image

* [x] add `readonly` property for form fields
* [x] allow to configure `readonly` in the properties panel
* [x] style `readonly` fields
* [ ] make it possible to add an expression to set the `readonly` condition
  * [ ] update `getSchemaVariables` 
* [x] add carbonized styles of `readonly`
* [x] update schema version
* [ ] update JSON schema (WIP)
* [ ] update linting rules
* [ ] update documentation

Up for discussion: Shortly, we potentially want to have a choice to provide an expression or a simple checkbox to define such a simple property (cf. https://github.com/bpmn-io/properties-panel/issues/194). 

__Why should we do it?__

Child of https://github.com/bpmn-io/form-js/issues/587
Depends on #374 
Design specs: https://www.figma.com/file/CiTSN1vNETjIAy6W9jUIFp/Form-Editor?node-id=1034-148176&t=T2sW2Ylfu9QonnHr-0

### PR DESCRIPTION
**Pull request by pinussilvestrus**
Closes #396 
Related to #653 

-----

Adds expression support for `readonly`.

Demo: https://demo-653-readonly-expressions--camunda-form-playground.netlify.app/

image

### CODE REVIEW

**Comment by pinussilvestrus** [comment]
We can't merge this until https://github.com/bpmn-io/properties-panel/pull/245 is done, but feel free to already have a look 👍

File: packages/form-js-viewer/src/render/hooks/useReadonly.js
Comment (vsgoulart) [inline packages/form-js-viewer/src/render/hooks/useReadonly.js:26]: Can we remove or create an issue for this?

**Review by vsgoulart — CHANGES_REQUESTED** [review]
Outine looks wrong for me on Firefox and Chroms

image

File: packages/form-js-viewer/src/render/hooks/useReadonly.js
Comment (pinussilvestrus) [reply packages/form-js-viewer/src/render/hooks/useReadonly.js]: I removed it.

**Comment by pinussilvestrus** [comment]
Seems like there were wrong styles applied. Could you try again?

**Comment by vsgoulart** [comment]
> Seems like there were wrong styles applied. Could you try again?

@pinussilvestrus Did you update the demo? The outline is still wrong for me

**Comment by vsgoulart** [comment]
@pinussilvestrus We also have some visual regression tests failing

**Comment by pinussilvestrus** [comment]
> @pinussilvestrus Did you update the demo? The outline is still wrong for me

It works for me on Firefox and Chrome (macOS).

image

> @pinussilvestrus We also have some visual regression tests failing

Yes, this is because we use another properties panel dependency which is not yet rebased on `v2.0.0`. Cf. https://github.com/bpmn-io/form-js/pull/657#issuecomment-1549709631. It will be fixed once we have the properties panel PR merged and integrated.

**Review by vsgoulart — APPROVED** [review]
@pinussilvestrus The focus is still wrong for me on Linux, maybe there's an issue because of the way we're consuming the properties panel dep? I'll approve if for now and I'll check it again when you use the real package
Can you please remind me of taking a look into this before you merge it?

**Comment by pinussilvestrus** [comment]
@vsgoulart the dependency is updated now. Can you maybe double check? For reference, the arrow action for opening a properties group should have the same outline color. If it's black on your machine as well, maybe it's a general issue we'd need to address in the properties panel.

image

**Comment by vsgoulart** [comment]
> @vsgoulart the dependency is updated now. Can you maybe double check? For reference, the arrow action for opening a properties group should have the same outline color. If it's black on your machine as well, maybe it's a general issue we'd need to address in the properties panel.
> 
> image

@pinussilvestrus Still wrong to me, but it seems unrelated to your PR so I just created a followup issue #664

**Comment by vsgoulart** [comment]
There's also some issues with the visual regression tests, I'll fix it tomorrow

**Review by vsgoulart — APPROVED** [review]

**Comment by pinussilvestrus** [comment]
@vsgoulart I updated the playwright version in the carbonization action as well (it's not on master, yet). I kept it in this PR.

**Comment by vsgoulart** [comment]
> Ok @vsgoulart seems like due to the updated playwright image the emojis are rendered differently sweat_smile
> 
> image
> 
> I guess we accept that? I'd update the screenshot then.

@pinussilvestrus yeah, let's just update the screenshot :+1:

**Comment by vsgoulart** [comment]
I think we should also generate a new `form-js` version and update it on Tasklist

**Review by christian-konrad — APPROVED** [review]
Works like a charm!

**Review by RomanKostka — APPROVED** [review]
Nice! Works as expected 🎉

i found a minor thing and am not sure if this was intended:

When i switch the toggle to `true`, then enable FEEL expressions, change the expression to `false` and then disable FEEL expressions the toggle is still `true`
image
image
I would expects that the toggle should become 'false'

The other way around it works as expected.
Switch the toggle to `true`, then again to `false`, enable FEEL expressions = `false`.

Also from the very beginning (before interacting with the toggle or the FEEL expression) the toggle is 'false' but the FEEL expression is empty.
image
I would expect the FEEL expression is `false` as the toggle is `false` too.

**Comment by christian-konrad** [comment]
@RomanKostka empty evaluates to `false` as well, so leaving it empty reduces the need to delete `false` first from the input, and users can immediately start typing. Hence I would propose to keep it as it is!

For detecting `false` for the toggle: I got your point. We could build this special logic, however we can never provide a smart detection of falsy values (e.g. an expression or variable evaluates to `false`) so I would say let's not spend time on this one case now.

**Pull request merged**
